primer/src/Primer/EvalFullInterp.hs

```diff
@@ -0,0 +1,444 @@
+--
+-- Warnings:
+-- - Trying to evaluate a divergent term will (unsurprisingly) not terminate,
+--   and may consume a lot of memory.
+-- - We 'throw' IO-based exceptions of type 'InterpError' for obvious issues
+--   in case expressions (though this will not happen with well-typed
+--   terms). This function will not throw 'Timeout' (that will only be
+--   thrown by the 'interp' wrapper).
+-- - To avoid these pitfalls, consider using 'interp', which limits runtime
+--   with 'timeout', and catches exceptions.
+--
+-- The interpreter is designed to be fairly simple, and to stay in the
+-- AST domain, but it does rely on subtle laziness properties (see below).
+-- We attempt to make it reasonably fast, given the constraint of simplicity,
+-- and maximally terminating.  We evaluate terms in some environment of local
+-- bindings, which enables us to share computation in @let x=<expensive>
+-- in f x x@ in a call-by-need fashion.  This is accomplished by the
+-- environment being a mapping of variable names to normal forms which
+-- are computed lazily.  The laziness is also used to avoid looping on
+-- diverging subterms which do not need to be fully-evaluated, for example
+-- in @head (letrec l=Cons True l in l)@, which should swiftly reduce
+-- to @True@. Another somewhat-common source of diverging-but-productive
+-- subterms are local recursive helper functions (since we evaluate under
+-- lambdas and inside let bindings).
+--
+-- We have made the design choice to use imprecise exceptions to be able to
+-- write simple code which can report obvious problems without compromising
+-- laziness. Another choice would be to not report errors (and letting these
+-- subterms be stuck). A non-option would be to use ExceptT to report errors,
+-- since that would destroy our critical laziness properties: the monadic
+-- bind would force the top Left/Right constructor, and that in turn would
+-- force the full evaluation of the relevent subterms. This design choice
+-- unfortunately precludes returning a partial result if we detect we have
+-- been evaluating for too long (it is possible to add a recursion depth
+-- limit and bail out if we hit it (though then evaluating recursive lets
+-- is less efficient), but we cannot easily then both report we timed out
+-- and also rebuild the whole term. Finding a smarter design that would
+-- allow this may be interesting future work.
+--
+-- The laziness property we rely on is that where possible we return the
+-- root node of an AST without forcing the evaluation of subterms. This
+-- makes it possible to pattern match on recursive calls to 'interp'' without
+-- incurring unbounded work. Note that this means we cannot make our 'Expr'
+-- type spine-strict without refactoring this interpreter!
+
+{-
+Future work:
+- can we reuse Redex module
+- perhaps do full NBE
+- maybe don't stay in AST domain
+  this perhaps may help with a pure recursion-depth limit and reporting partial results?
+-}
```

Comment:
We should expand on this, maybe in a GitHub issue.

Reply:
See #1194